### PR TITLE
set_fill(), set_color() and set_stroke() improvements

### PR DESCRIPTION
**Pull request by sparshg**
## Motivation

The three methods seem to create some confusion, even in the failing test, set_color() is used to change outline colors, but set_stroke() should be used to do that. So, its not really a breaking change, just making things more intuitive.
### Before:
**set_fill(), set_stroke()** - Changes fill/stroke color, doesn't affect opacity if not mentioned
**set_color()** - just calls the above two methods together
### Problem:
The default fill color has 0 opacity, as a result doing this:
```py
Circle().set_stroke(BLUE) - a BLUE outlined circle
Circle().set_fill(BLUE) - a default colored outlined circle
Circle().set_color(BLUE) - a BLUE outlined circle #changes fill and stroke color only
```
Isn't that confusing because the default fill is transparent?
### After:
With new change, "if no opacity is mentioned, it will be set to 1 if it was 0" for set_fill() and set_stroke().
set_color() works the same. Also added optional opacity parameter for set_color()
```py
Circle().set_stroke(BLUE) - a BLUE outlined circle
Circle().set_fill(BLUE) - a default colored outlined + blue filled circle
Circle().set_color(BLUE) - a BLUE outlined circle #changes fill and stroke color only
Circle().set_color(BLUE, opacity=1) - a BLUE outlined circle + filled circle #sets opacity for both fill and stroke
```
And this looks much more intuitive and better. 
Also set_color_only() function is added which doesn't affect these opacities, (to make FadeToColor animation work as intended)
Added docs to some functions

## Oneline Summary of Changes

```
- set_fill() and set_stroke() improved, set_color() optional opacity parameter added (:pr:`956`)
```

## Acknowledgements
- [x] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)

## Reviewer Checklist
- [ ] Newly added functions/classes are either private or have a docstring
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings
- [ ] The oneline summary has been included [in the wiki](https://github.com/ManimCommunity/manim/wiki/Changelog-for-next-release)

### CODE REVIEW

**Comment by sparshg** [comment]
okayy I guess I will have to do some changes

**Review by huguesdevimeux — DISMISSED** [review]
SGTM! 

Do you think you could add a graphical unit test for that ? Everything is explained there. 
https://github.com/ManimCommunity/manim/wiki/Testing

If you're not sure or you don't understand, please tell !

**Comment by huguesdevimeux** [comment]
EDIT : I was too fast and didn't see the failing tests. 

  I believe failing tests are normal as IIRC the failing test is the one testing set_sheen. Then, you'll have to rewrite the failing test and to regenerate the control data. cf [wiki](https://github.com/ManimCommunity/manim/wiki/Testing). 

If you're not sure, feel free to tell and I can do it for you.

**Comment by sparshg** [comment]
Yeah I am doing some changes and will comment explaining the things and adding some docs

**Comment by sparshg** [comment]
Actually whats failing is the set_color() thing. According to me these three functions, set_color(), set_stroke(), set_fill() seem to overlap and make confusion.

For example I make a Square().set_color(BLUE). Now, according to code set_color() sets both fill + stroke. But you only see that the stroke gets changed, (because the fill is transparent by default... I mean if a person doesn't check the code how can you know that set_color doesn't just fill the outlines.... and then set_fill seems to do nothing because fill is transparent.. so you have to use opacity parameter which is again confusing if you don't check the code... its so unintuitive to me).

Similarly, set_fill() is used to fill the object and make it opaque if it was transparent to avoid confusion and avoiding the need to do set_color(BLUE).set_opacity(1) or doing set_fill(BLUE, opacity=1). Added similar thing to set_stroke()

**Added explanation to docs too.**

**In the failing test**
set_color() is used to make the outlines RED in output. But my change will make the square fill also red. If one wants to set the stroke, then they should use set_stroke(), this will avoid such confusions. So its intended change.

**Comment by huguesdevimeux** [comment]
@sparshg I just fixed and added some new tests, so you don't have to worry about it. 

If you're good, then I'll approve!

**Comment by sparshg** [comment]
@hugesdevimeux looks good imo

**Review by huguesdevimeux — APPROVED** [review]

File: manim/mobject/types/vectorized_mobject.py [hunk 1]
```diff
@@ -367,6 +425,23 @@ def get_color(self):
         return self.get_fill_color()
 
     def set_sheen_direction(self, direction, family=True):
+        """Changes the direction of the sheen applied, use set_sheen() to apply
```
Comment (behackl) [inline manim/mobject/types/vectorized_mobject.py:428]:
```suggestion
        """Changes the direction of the applied sheen.
```

File: manim/mobject/types/vectorized_mobject.py [hunk 2]
```diff
@@ -367,6 +425,23 @@ def get_color(self):
...
+
+        Parameters
+        --------
```
Comment (behackl) [inline manim/mobject/types/vectorized_mobject.py:431]:
```suggestion
        ----------
```

File: manim/mobject/types/vectorized_mobject.py [hunk 3]
```diff
@@ -367,6 +425,23 @@ def get_color(self):
...
+
+        Examples
+        --------
+        Normal usage::
+
+            Circle().set_sheen_direction(UP)
+
+        See Also
+        --------
+        :meth:`set_sheen`
```
Comment (behackl) [inline manim/mobject/types/vectorized_mobject.py:443]:
```suggestion
        :meth:`~.VMobject.set_sheen`
```

File: manim/mobject/types/vectorized_mobject.py [hunk 4]
```diff
@@ -376,6 +451,26 @@ def set_sheen_direction(self, direction, family=True):
         return self
 
     def set_sheen(self, factor, direction=None, family=True):
+        """Applies a gradient from a direction to introduce a soft lustre.
```
Comment (behackl) [inline manim/mobject/types/vectorized_mobject.py:454]:
```suggestion
        """Applies a color gradient from a direction.
```

File: manim/mobject/types/vectorized_mobject.py [hunk 5]
```diff
@@ -376,6 +451,26 @@ def set_sheen_direction(self, direction, family=True):
...
+
+        Parameters
+        --------
+        factor : :class:`float`
+            The extent of lustre/gradient to apply. If negative, the gradient starts from black, \
+            if positive the gradient starts from white (to current color)
```
Comment (behackl) [inline manim/mobject/types/vectorized_mobject.py:460]:
```suggestion
        factor : :class:`float`
            The extent of lustre/gradient to apply. If negative, the gradient
            starts from black, if positive the gradient starts from white and
            changes to the current color.
```

File: manim/mobject/types/vectorized_mobject.py [hunk 6]
```diff
@@ -376,6 +451,26 @@ def set_sheen_direction(self, direction, family=True):
...
+        direction : :class:`numpy.ndarray`, optional
+            Direction from where the gradient is applied
```
Comment (behackl) [inline manim/mobject/types/vectorized_mobject.py:462]:
```suggestion
            Direction from where the gradient is applied.
```

File: manim/mobject/types/vectorized_mobject.py [hunk 7]
```diff
@@ -376,6 +451,26 @@ def set_sheen_direction(self, direction, family=True):
...
+           
+        Examples
+        ---------
```
Comment (behackl) [inline manim/mobject/types/vectorized_mobject.py:465]:
```suggestion
        --------
```

**Review by behackl — CHANGES_REQUESTED** [review]
The changes in the code look good to me; the documentation needs some more work though. I've added a bunch of suggestions.

**Comment by behackl** [comment]
... and thank you for your contribution, of course! It is much appreciated!

File: manim/mobject/types/vectorized_mobject.py [hunk 8]
```diff
@@ -275,9 +309,33 @@ def match_style(self, vmobject, family=True):
...
+            Color for fill and stroke
+        opacity : :class:`float`, optional
+            Opacity to fill and stroke
+        """
+        self.set_fill(color, opacity, family=family)
+        self.set_stroke(color, opacity, family=family)
+        self.color = colour.Color(color)
+        return self
+
+    def set_color_only(self, color, family=True):
```
[the remark this answers is not in the document]
Comment (sparshg) [reply manim/mobject/types/vectorized_mobject.py:333]: idk if there is a better solution, for FadeToColor animation to work as before, the opacities of fill and color should remain unaltered. (btw I fixed this function, forgot to remove the testing code)

**Comment by sparshg** [comment]
I guess everything is fixed now, sorry for so many commits lol that was my first pull request...

**Review by huguesdevimeux — APPROVED** [review]
Tests SGTM. 

Waiting for @behackl review and we can merge!

**Comment by sparshg** [comment]
The conflict arose due to my own PR #1003, should try to fix

**Comment by eulertour** [comment]
Color and opacity shouldn't be coupled, the user should call set_stroke/fill_opacity if they want opacity to change.

**Comment by sparshg** [comment]
@eulertour I thought that setting opacity with color would make sense because set_color() calls set_fill() and set_stroke()
So set_color(opacity) now calls set_fill(opacity), set_stroke(opacity)
And for that purpose I did create set_color_only() function which doesn't set opacity

**Comment by eulertour** [comment]
Well setting color without the opacity should be the default for a method called set_color. set_color should still call set_fill and set_stroke to change the color for each of them, just not the opacity.

**Comment by sparshg** [comment]
wait, but opacity parameter is optional for set_color() soo without opacity it works the same by just calling set_fill() and set_stroke().

**Comment by eulertour** [comment]
If calling set_color without opacity doesn't change the opacity then that's fine, but the description says it'll set opacity to 1 if opacity isn't passed.

If the user specifies an opacity in set_color then it makes sense to update opacity of course.

**Comment by sparshg** [comment]
yeah because I changed set_fill() and set_stroke() to do that and set_color() calls those 2 so thats why opacity is changing.
So you mean set_color() should not do this and retain original behaviour (which set_color_only does)

**Comment by eulertour** [comment]
Yes, set_color should only affect color by default.

**Comment by sparshg** [comment]
@eulertour ok as you said, set_fill and set_stroke will change opacity to 1 if 0 and nothing passed
but set_color() works like before with an optional opacity parameter

**Comment by eulertour** [comment]
Nothing should set opacity to 1 if it's 0 by default, opacity should only change if the user calls for it to be changed.

**Comment by sparshg** [comment]
uhh then what was the point of this PR lol, this PR sets fill/stroke color to 1 if it was 0
like Circle().set_fill(BLUE) and Circle.set_stroke(BLUE) currently have same output coz of transparent fill which this change fixes.... Other things are same as before

**Comment by eulertour** [comment]
I don't think this change should be made at all, it presupposes too much about what the user wants to do.

**Comment by sparshg** [comment]
ok whatever, I just wanted this change coz when I started using manim, I spent ages knowing that the default fill opacity of shapes is transparent, and setting color/fill doesn't do anything lol. I still think these 3 methods should change opacity to 1 if was 0 and nothing is mentioned.

**Comment by eulertour** [comment]
> when I started using manim, setting color/fill doesn't do anything

That shouldn't happen, if you set the color when the stroke opacity isn't 0 that should be visible.

**Comment by sparshg** [comment]
I meant this: This won't do anything
```py
circle = Circle()
self.add(circle)
self.play(circle.animate.set_fill(BLUE))
```
Then I tried using set_stroke and set_color which changed outline only... Soo I had to watch some tutorial to clear the confusion

**Comment by eulertour** [comment]
Yeah, you'd have to also update the fill opacity to see that, since the attributes aren't coupled.

**Comment by PhilippImhof** [comment]
I feel there is some dilemma here. On one hand, `set_fill` should not mess with the opacity that the user might have set earlier. On the other hand, 
```python
circle = Circle()
self.add(circle)
self.play(circle.animate.set_fill(BLUE))
```
should definitely result in that circle being blue...

**Comment by eulertour** [comment]
I disagree, the method should just do what the name indicates.

You made a set_color_only() method earlier, so why not add a set_opaque_color() or set_color_with_with_opacity() method?

**Comment by sparshg** [comment]
or.. why don't we just change the default fill value to opaque instead like every application does... This way we won't have to do these changes and if one wants to make a shape fill transparent they can do Circle(fill_opacity=0) or Circle().set_fill(opacity=0).

**Comment by eulertour** [comment]
The shapes with the fill are generally more useful for illustrating concepts than ones with fill, so that's why it's 0 be default.

**Comment by PhilippImhof** [comment]
> I disagree, the method should just do what the name indicates.

Absolutely. But it should do it in a reasonable way. If the function sets the fill color, but leaves it fully transparent, that's most certainly not what one expects. 

If the opacity is currently 0, it could be set to 1, but an optional parameter could be set to inhibit this behaviour for those who would really want to set a fill color without it being visible.

**Comment by sparshg** [comment]
>  but an optional parameter could be set to inhibit this behaviour for those who would really want to set a fill color without it being visible.

That can already be done by just saying set_fill(opacity=0) though

**Comment by PhilippImhof** [comment]
> That can already be done by just saying set_fill(opacity=0) though

Yes, but I thought something like `do_not_touch_my_very_well_chosen_opacity` would make the idea clearer. The user might not want to restate their value.

**Comment by sparshg** [comment]
imo 0 default opacity doesn't make sense, it should be "None" instead of 0, and now we can check if set_fill() is called and `current opacity == None and passed opacity == None` then set opacity to 1.. so this might be another solution

**Comment by sparshg** [comment]
> Yes, but I thought something like `do_not_touch_my_very_well_chosen_opacity` would make the idea clearer. The user might not want to restate their value.

If you think, this situation will never arise...
The change will only alter opacity if opacity is 0 and set_fill is called... (why would you call set_fill with 0 opacity -_-)
or if opacity is 0 and set_stroke is called... (why would you call set_stroke with 0 opacity -_-)

I don't see a reason to deny these changes actually

**Comment by PhilippImhof** [comment]
> What I described before is the most reasonable way. 

Setting the fill color without it being visible seems all but reasonable to me, but that is just personal.

**Comment by eulertour** [comment]
The fill color is only an attribute. If you add or move points on a mobject with 0 opacity, there's no reason for that to update the opacity too.

If you want an animation to behave that way then that would make more sense since their primary purpose is to be played, but attributes are just for storing data.

**Comment by sparshg** [comment]
That code was just for example... the problem is simply that this produces a red outlined circle:
'''py
circle=Circle()
#do stuff
circle.set_fill(BLUE)
self.wait()
'''

**Comment by eulertour** [comment]
It looks like it wouldn't produce anything since the circle wasn't added to the scene, but in any case that's expected behavior since the default fill opacity is 0.

**Pull request closed without merging**